src/Processors/Transforms/DeduplicationTokenTransforms.h

```diff
@@ -0,0 +1,202 @@
+#pragma once
+
+#include <Processors/Chunk.h>
+#include <Processors/ISimpleTransform.h>
+
+#include <base/defines.h>
+
+
+namespace DB
+{
+    class RestoreChunkInfosTransform : public ISimpleTransform
+    {
+    public:
+        RestoreChunkInfosTransform(Chunk::ChunkInfoCollection chunk_infos_, const Block & header_)
+                : ISimpleTransform(header_, header_, true)
+                , chunk_infos(std::move(chunk_infos_))
+        {}
+
+        String getName() const override { return "RestoreChunkInfosTransform"; }
+
+        void transform(Chunk & chunk) override;
+
+    private:
+        Chunk::ChunkInfoCollection chunk_infos;
+    };
+
+
+namespace DeduplicationToken
+{
+    class TokenInfo : public ChunkInfoCloneable<TokenInfo>
+    {
+    public:
+        TokenInfo() = default;
+        TokenInfo(const TokenInfo & other) = default;
+
+        String getToken(bool enable_assert = true) const;
+
+        bool empty() const { return parts.empty(); }
+        bool tokenInitialized() const { return stage != INITIAL && stage != SOURCE_BLOCK_NUMBER; }
+
+        void addPieceToInitialToken(String part);
+        void closeInitialToken();
+        void setUserToken(const String & token);
+        void setSourceBlockNumber(size_t sbn);
+        void setViewID(const String & id);
+        void setViewBlockNumber(size_t mvbn);
+        void reset();
+
+    private:
+        void addTokenPart(String part);
```

Comment:
May be by reference?

Reply:
No. In that way it works efficiently with both types of ards: const ref and r-value.
If we took here const ref, it would make copies for all r-value args.
const ref is a common approach when you do not alter arg and do not copy it.
Otherwise temp object is better. It is like const ref + r-value. Use it when you do not have to restrict usage only with one form.